### PR TITLE
Leap approaches

### PR DESCRIPTION
**Pull request by bobahop**

### CODE REVIEW

**Comment by bobahop** [comment]
Initial commits for file structure and code.

**Comment by bobahop** [comment]
@ErikSchierboom Leap approaches  is ready for review. I'm sorry, there may be indentation inconsistencies that may have originated from different file types (e.g. md, txt and cs.)

File: exercises/practice/leap/.approaches/built-in-method/content.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# `DateTime.IsLeapYear`
+
+```csharp
+using System; // for DateTime
+
+public static bool IsLeapYear(int year)
+{
+  DateTime.IsLeapYear(year);
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/built-in-method/content.md:8]:
```suggestion
    return DateTime.IsLeapYear(year);
```

File: exercises/practice/leap/.approaches/boolean-chain/snippet.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+public static bool IsLeapYear(int year)
+{
+        return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/boolean-chain/snippet.txt:3]:
```suggestion
    return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```

File: exercises/practice/leap/.approaches/boolean-chain/content.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+The chain of boolean expressions is efficient, as it proceeds from testing the most likely to least likely conditions.
+
+## Shortening
+
+When the body of a function is a single expression, the function can be implemented as an [expression-bodied member][expression-bodied-member], like so
+
+```csharp
+public static bool IsLeapYear(int year) =>
+        year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/boolean-chain/content.md:33]:
```suggestion
    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```

File: exercises/practice/leap/.approaches/boolean-chain/content.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+# Chain of boolean expressions
+
+```csharp
+public static bool IsLeapYear(int year)
+{
+        return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/boolean-chain/content.md:6]:
```suggestion
    return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```

File: exercises/practice/leap/.approaches/boolean-chain/content.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+- If the year is not evenly divisible by `100`, then the expression is `true` and the chain will "short-circuit" to return `true`,
+since the next operator is a [logical OR][logical-or] (`||`).
+- If the year _is_ evenly divisible by `100`, then the expression is `false`, and the returned value from the chain will be if the year is evenly divisible by `400`.
+
+| year | year % 4 == 0 | year % 100 != 0 | year % 400 == 0 | is leap year |
+| ---- | ------------- | --------------- | --------------- | ------------ |
+| 2020 |          true |            true |   not evaluated |         true |
+| 2019 |         false |   not evaluated |   not evaluated |        false |
+| 2000 |          true |           false |            true |         true |
+| 1900 |          true |           false |           false |        false |
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/boolean-chain/content.md:22]: Lovely!

File: exercises/practice/leap/.approaches/built-in-method/content.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+This may be considered a "wicked cheat" for this exercise, by simply passing the year into the [IsLeapYear][is-leap-year] method of the [DateTime][datetime] struct.
+Although it is not in the spirit of this exercise, `IsLeapYear` would be the idiomatic way to determine if a year is a leap year in the "real world".
+
+## Shortening
+
+When the body of a function is a single expression, the function can be implemented as an [expression-bodied member][expression-bodied-member], like so
+
+```csharp
+public static bool IsLeapYear(int year) =>
+  DateTime.IsLeapYear(year);
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/built-in-method/content.md:21]:
```suggestion
    DateTime.IsLeapYear(year);
```

File: exercises/practice/leap/.approaches/built-in-method/snippet.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+public static bool IsLeapYear(int year)
+{
+  return DateTime.IsLeapYear(year);
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/built-in-method/snippet.txt:3]:
```suggestion
    return DateTime.IsLeapYear(year);
```

File: exercises/practice/leap/.approaches/datetime-addition/snippet.txt [hunk 8]
```diff
@@ -0,0 +1,4 @@
+public static bool IsLeapYear(int year)
+{
+  return (new DateTime(year, 2, 28)).AddDays(1.0).Day == 29;
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/datetime-addition/snippet.txt:3]:
```suggestion
    return (new DateTime(year, 2, 28)).AddDays(1.0).Day == 29;
```

Comment (ErikSchierboom) [reply exercises/practice/leap/.approaches/datetime-addition/snippet.txt:3]: Side note: I find it annoying that `AddDays` takes a `double`, not an `int`.

File: exercises/practice/leap/.approaches/introduction.md [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+```
+
+For more information, check the [Boolean chain approach][approach-boolean-chain].
+
+## Approach: Ternary operator of Boolean expressions
+
+```csharp
+public static bool IsLeapYear(int year)
+{
+        return year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/introduction.md:29]:
```suggestion
    return year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```

File: exercises/practice/leap/.approaches/introduction.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+The key to solving  Leap is to know if the year is divisible by `4`, `100` and `400`.
+For determining that, you will use the the [remainder operator][remainder-operator].
+
+## Approach: Chain of Boolean expressions
+
+```csharp
+public static bool IsLeapYear(int year)
+{
+        return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/introduction.md:18]:
```suggestion
    return year % 4 == 0 && (year % 100 != 0 || year % 400 == 0);
```

File: exercises/practice/leap/.approaches/ternary-operator/content.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
+# Ternary operator
+
+```csharp
+public static bool IsLeapYear(int year)
+{
+  return year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/ternary-operator/content.md:6]:
```suggestion
    return year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```

File: exercises/practice/leap/.approaches/ternary-operator/content.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+making it less efficient than another approach that would first test if the year is evenly divisible by `4`,
+which is more likely than the year being evenly divisible by `100`.
+
+## Shortening
+
+When the body of a function is a single expression, the function can be implemented as an [expression-bodied member][expression-bodied-member], like so
+
+```csharp
+public static bool IsLeapYear(int year) =>
+  year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/ternary-operator/content.md:35]:
```suggestion
    year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```

File: exercises/practice/leap/.approaches/ternary-operator/snippet.txt [hunk 13]
```diff
@@ -0,0 +1,4 @@
+public static bool IsLeapYear(int year)
+{
+  return year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```
Comment (ErikSchierboom) [inline exercises/practice/leap/.approaches/ternary-operator/snippet.txt:3]:
```suggestion
    return year % 100 == 0 ? year % 400 == 0 : year % 4 == 0;
```

**Review by ErikSchierboom — APPROVED** [review]

**Comment by bobahop** [comment]
@ErikSchierboom  Thank you very much for eyeballing all of those pesky indentation inconsistencies. I'll take that burden from now on...

All checks have passed. Should be ready for merge unless you want to check it over again.

**Comment by ErikSchierboom** [comment]
Thanks!

**Pull request merged**